### PR TITLE
changes to review

### PR DESCRIPTION
**Pull request by minhtridn2001**
# Basic review checklist
## 1. Architecture & Design
- [x] Are layers properly separated (Controller, Service, Repository, etc.)?
  - Suggestions:
    - Rename files like UserServiceImp -> UserServiceImpl (Impl is a widely accepted abbreviation)
    - Reorganize the package as below to follow the common convention:
      - Before
      ```
          service/
            Interface/
              ProductService.java
            Implement/
              ProductServiceImpl.java
      ```
      - After
      ```
          service/
            impl/
              ProductServiceImpl.java
            ProductService.java
      ```
    - Your current organization structure is fine, but switching to subdomains structure can improve modularity, scalability, and maintainability. 
      - Before
      ```
      com.example.apisample
        config/
        controller/
          UserController.java
          ProductController.java
        entity/
          Product.java
          User.java
        exception/
          productservice/
            ProductDeletedException.java
          userservice/
            UserDoesNotExistException.java
      ```
      - After
      ```
      com.example.apisample
        config/
        user/
          controller/
            UserController.java
          entity/
            User.java
          exception/
            UserDoesNotExistException.java
        product/
          controller/
            ProductController.java
          entity/
            Product.java
          exception/
            ProductDeletedException.java
      ```
- [x] Are DTOs used to avoid exposing JPA entities in the API?
- [x] Is Dependency Injection used instead of manual instantiation?
- [ ] Is business logic kept inside the Service layer?
  - Avoid putting business logic in, Controllers, DTOs, e.g., `UserRegisterRequestDTO`, `AuthController`.

---

## 2. REST API Design
- [ ] Are the endpoints RESTful and clearly defined? (GET, POST, PUT, DELETE)
  - URIs should be resource-based, not method-based, e.g.,
    - GET /products/get-all   -> Should be just GET /products 
    - GET /products/get/{id}  -> Should be just GET /products/{id}
- [x] Are appropriate HTTP status codes used? (200, 201, 400, 401, 403, 500, etc.)
- [ ] Is input validation implemented? (`@Valid`, `@NotNull`, `@Size`, etc.)
  - Missing input validation.
---

## 3. Entity & Database Mapping
- [x] Is the entity-to-table mapping accurate?
- [x] Are primary keys defined correctly in entities?
- [ ] Are appropriate data types used for each field?
  - Remove @Lob for Category.description
- [x] Are relationships (`@ManyToOne`, `@OneToMany`, etc.) clearly defined?
- [x] Are fetch types used appropriately? (EAGER vs LAZY)
- [ ] Are cascade types used correctly? (e.g., `CascadeType.ALL`, `CascadeType.PERSIST`)
- [ ] Are `@Column` annotations used for custom column names and constraints?
  - `User.email` should be unique, and not nullable`

Suggestions:
- Use @CreationTimestamp and @UpdateTimestamp annotations for automatic timestamp management. 
- The `User` entity implements `UserDetails` directly, which tightly couples User domain model to Spring Security. Consider using a separate class for Spring Security integration
- Consider adding version fields with @Version for optimistic locking where appropriate, e.g., `Product` entity.
---

## 4. Repository Layer
- [x] Are methods named clearly and do they follow naming conventions?
- [x] Are `@Query` or `CriteriaBuilder` used correctly?
- [x] Are custom queries too complex or hard to read?
- [x] Are pagination and sorting implemented in queries where appropriate?

---

## 5. Service Layer
- [ ] Are classes violating separation of concerns?
  - E.g., `UserServiceImp` is doing too much. Consider splitting user management, authentication.
- [ ] Are exceptions handled properly? (try-catch, custom exceptions)
  - Use unchecked exceptions for programming, validation, business logic errors (e.g., `UserDoesNotExistException`, `ProductNotFoundException`, etc.).
  - Use checked exceptions recoverable error such as external system errors (e.g., database, network).
- [ ] Are transactions managed correctly? (`@Transactional`)
  - Add @Transactional for methods that modify data.

Others:
- Add @Override annotation for interface methods in `UserServiceImp` and `ProductServiceImp`.
- `UserServiceImp` should not implement `UserDetailsService`. Consider creating a separate class for authentication.
---

## 6. Controller Layer
- [x] Does the controller only handle request/response and delegate to services?
- [x] Are annotations like `@GetMapping`, `@PostMapping`, etc. used correctly?
- [x] Is pagination (`Pageable`) and sorting implemented where appropriate?

---

## 7. Exception Handling
- [x] Is there a global exception handler using `@ControllerAdvice`?
- [ ] Are exceptions meaningful and helpful for debugging?
  - Your custom exceptions provide basic error signaling but lack detailed information.
- [ ] Are errors properly logged?
  - Miss logging for exceptions in many places.

---

## 8. Unit Testing
- [ ] Are there tests for core service methods?
- [ ] Are repository methods with custom queries covered?
- [ ] Are tests clear and easy to understand?
- [ ] Are edge cases and error scenarios tested?
---

## 9. Logging & Debugging
- [ ] Is the logging level (ERROR, WARN, INFO, DEBUG) used appropriately?
  - Use INFO for general information, WARN for potential issues, and ERROR for actual errors.
  - Use DEBUG for detailed information during development.
- [ ] Are logs clear and helpful for tracing issues?
- [ ] Are exception stack traces logged where needed (log.error("...", exception))?
  - Bad: logger.error("Error occurred");
  - Good: logger.error("Error occurred", ex);
- [x] Are sensitive data (e.g., passwords, tokens) excluded from logs?

Suggestions:
- Most of your INFO logs should be DEBUG, which should only be output during development. Otherwise, it will cause too many unnecessary logs in production.
---

## 10. Clean Code & Best Practices
- [x] Are variable and method names meaningful?
- [x] Are methods concise? Are there any long methods that could be split into smaller ones?
- [ ] Is there duplicated logic that could be refactored?
  - E.g., in `UserServiceImp`, user retrieval and validation is repeated frequently.
- [x] Is `Optional` used instead of returning `null`?
  - In some places, e.g., in `UserServiceImp`, null checks are done instead of using `Optional`.
- [ ] Is the `Stream API` used appropriately to improve conciseness and readability?
- [ ] Are there any hardcoded values in the code?

---

## 11. Basic Security
- [x] Are sensitive endpoints secured?
- [x] Are passwords securely hashed?
- [x] Are sensitive data (e.g., passwords, tokens) excluded from logs?
- [x] Are there any hardcoded secrets in the code?

### CODE REVIEW

File: Api Sample/src/main/java/com/example/apisample/model/dto/message/LogMessage.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package com.example.apisample.model.dto.message;
+
+
+public class LogMessage {
+    //Auth Controller Log
+    public static final String logSuccessRefresh = "Token has been refreshed";
```
Comment (minhtridn2001) [inline Api Sample/src/main/java/com/example/apisample/model/dto/message/LogMessage.java:6]: For constants in Java, the standard naming convention is SCREAMING_SNAKE_CASE
`logSuccessRefresh` -> `LOG_SUCCESS_REFRESH`

File: Api Sample/src/main/java/com/example/apisample/model/dto/message/LogMessage.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public static final String logSuccessUploadImage = "User profile image upload successfully!";
+    public static final String logStartGetUserByEmail = "Getting user by email";
+    public static final String logSuccessGetUserByEmail = "User get successfully!";
+    public static final String logStartGetUserById = "Getting user by id";
+    public static final String logSuccessGetUserById = "User get successfully!";
+    //Email Log
+    public static final String logEmailSendFailed = "Email sending failed";
+
+    //Role Log
+    public static String logStartGetAllRoles = "Start to get all roles";
```
Comment (minhtridn2001) [inline Api Sample/src/main/java/com/example/apisample/model/dto/message/LogMessage.java:47]: While using constants is a common practice for log messages, there are a few considerations:
Benefits of using constants:
- Consistency in log messages
- Centralized management of messages
- Easier to search for specific log patterns
- Better for internationalization if needed later

Drawbacks:
- Adds verbosity to the code
- Creates many string constants that might be used only once
- Makes the log messages less contextual

For simple logging scenarios, inline strings can be more readable and maintainable than numerous constants.
